client/dashboard/customizable.js

```diff
@@ -4,56 +4,79 @@
  */
 import { __ } from '@wordpress/i18n';
 import { Component, Fragment } from '@wordpress/element';
+import { compose } from '@wordpress/compose';
 import { applyFilters } from '@wordpress/hooks';
+import { withDispatch } from '@wordpress/data';
+
+/**
+ * WooCommerce dependencies
+ */
+import { H, ReportFilters } from '@woocommerce/components';
 
 /**
  * Internal dependencies
  */
 import './style.scss';
-import DashboardCharts from './dashboard-charts';
-import Leaderboards from './leaderboards';
+import sectionsDefinition from './sections-definition';
 import Section from './section';
-import { ReportFilters, H } from '@woocommerce/components';
-import StorePerformance from './store-performance';
+import withSelect from 'wc-api/with-select';
 
-// @todo Replace dashboard-charts, leaderboards, and store-performance sections as neccessary with customizable equivalents.
-export default class CustomizableDashboard extends Component {
+class CustomizableDashboard extends Component {
 	constructor( props ) {
 		super( props );
+
 		this.state = {
-			sections: applyFilters( 'woocommerce_dashboard_sections', [
-				{
-					key: 'store-performance',
-					component: StorePerformance,
-					title: __( 'Store Performance', 'woocommerce-admin' ),
-				},
-				{
-					key: 'charts',
-					component: DashboardCharts,
-					title: __( 'Charts', 'woocommerce-admin' ),
-				},
-				{
-					key: 'leaderboards',
-					component: Leaderboards,
-					title: __( 'Leaderboards', 'woocommerce-admin' ),
-				},
-			] ),
+			sections: this.mergeSections(
+				applyFilters( 'woocommerce_dashboard_sections', sectionsDefinition ),
+				props.userPrefSections || []
+			),
+		};
+
+		this.updateSection = this.updateSection.bind( this );
+	}
+
+	mergeSections( defaultSections, prefSections ) {
+		const defaultKeys = defaultSections.map( section => section.key );
+		const prefKeys = prefSections.map( section => section.key );
+		const keys = new Set( [ ...prefKeys, ...defaultKeys ] );
+		const sections = [];
+
+		keys.forEach( key => {
+			const prefSection = prefSections.find( section => section.key === key );
+			const defaultSection = defaultSections.find( section => section.key === key );
```

Comment:
Yes, that's to be expected. When you hadn't done any modification to the _Dashboard_, `prefSections` will be empty but that doesn't cause an error when merging the section.

I [added a check](b5fe9cdf1db6ffde39188bb80c20bd59ac1cdf59) to make sure `defaultSection` is defined, though. Since that one is required because it contains the information of which component to use.